File: cpp/WeightedUnionFind.cpp
template<typename T>
struct WeightedUnionFind{
    vector<int> par;
    vector<T> weight;
    vector<bool> nan;
    WeightedUnionFind(int n):par(n, -1), weight(n,0), nan(n, false){}

    int root(int x){
        if(par[x] < 0){
            return x;
        }
        int r = root(par[x]);
        weight[x] += weight[par[x]];
        return par[x] = r;
    }

    T get_w(int x){
        root(x);
        return weight[x];
    }

    int size(int x) {
        return -par[root(x)];
    }

    T diff(int x, int y){
        return get_w(y) - get_w(x);
    }

    bool is_nan(int x){
        return nan[root(x)];
    }

    void merge(int x, int y, T w){
        //w[y] - w[x] = w

        w += get_w(x);
        w -= get_w(y);

        x = root(x);
        y = root(y);

        if(x == y){
            if(w != 0){
                nan[x] = true;
            }
            return;
        }

        if(par[x] > par[y]){
            swap(x, y);
            w = -w;
        }

        par[x] += par[y];
        par[y] = x;
        weight[y] = w;
    }

    bool same(int x, int y){
        return root(x) == root(y);
    }
};

File: rust/WeightedUnionFind.rs
struct WeightedUnionFind {
    par: Vec<i32>,
    weight: Vec<i32>,
}

#[allow(dead_code)]
impl WeightedUnionFind {
    fn new(n: usize) -> WeightedUnionFind {
        WeightedUnionFind {
            par: vec![-1; n],
            weight: vec![0; n],
        }
    }

    fn root(&mut self, x: usize) -> usize{
        if self.par[x] < 0 {
            return  x;
        }

        let p = self.root(self.par[x] as usize);
        self.weight[x] += self.weight[self.par[x] as usize];
        self.par[x] = p as i32;
        p
    }

    fn get_w(&mut self, x: usize) -> i32{
        self.root(x);
        self.weight[x]
    }

    fn diff(&mut self, x: usize, y: usize) -> i32{
        self.get_w(y) - self.get_w(x)
    }

    fn size(&mut self, x: usize) -> i32 {
        let root = self.root(x);
        -self.par[root]
    }

    fn merge(&mut self, x: usize, y: usize, mut w: i32) {
        w += self.get_w(x);
        w -= self.get_w(y);

        let mut x = self.root(x);
        let mut y = self.root(y);

        if x == y {
            return;
        }

        if self.par[x] > self.par[y] {
            std::mem::swap(&mut x, &mut y);
            w = -w;
        }

        self.par[x] += self.par[y];
        self.par[y] = x as i32;
        self.weight[y] = w;
    }

    fn same(&mut self, x: usize, y: usize) -> bool {
        self.root(x) == self.root(y)
    }
}
